src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -182,191 +184,220 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
+	pendingTransactionList = PendingWorkerTransactionList(connection);
+	pendingTransactionSet = ListToHashSet(pendingTransactionList, NAMEDATALEN, true);
 
-	return recoveredTransactionCount;
-}
+	/* find in-progress distributed transactions */
+	activeTransactionNumberList = ActiveDistributedTransactionNumbers();
+	activeTransactionNumberSet = ListToHashSet(activeTransactionNumberList,
+											   sizeof(uint64), false);
 
+	/* scan through all recovery records of the current worker */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_transaction_groupid,
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(groupId));
 
-/*
- * NameListDifference returns the difference between the bag of
- * names in nameList and subtractList. Both are assumed to be
- * sorted. We cannot use list_difference_ptr here since we need
- * to compare the actual strings.
- */
-static List *
-NameListDifference(List *nameList, List *subtractList)
-{
-	List *differenceList = NIL;
-	ListCell *nameCell = NULL;
+	/* get a snapshot of pg_dist_transaction */
+	scanDescriptor = systable_beginscan(pgDistTransaction,
+										DistTransactionGroupIndexId(), indexOK,
+										NULL, scanKeyCount, scanKey);
 
-	int subtractIndex = 0;
-	int subtractCount = list_length(subtractList);
-	char **subtractArray = (char **) PointerArrayFromList(subtractList);
+	/* find stale prepared transactions on the remote node */
+	recheckTransactionList = PendingWorkerTransactionList(connection);
+	recheckTransactionSet = ListToHashSet(recheckTransactionList, NAMEDATALEN, true);
 
-	foreach(nameCell, nameList)
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
 	{
-		char *baseName = (char *) lfirst(nameCell);
+		bool isNull = false;
+		bool isTransactionInProgress = false;
+		bool isTransactionPending = false;
 
-		bool nameFound = FindMatchingName(subtractArray, subtractCount,
-										  baseName, &subtractIndex);
+		Datum transactionNameDatum = heap_getattr(heapTuple,
+												  Anum_pg_dist_transaction_gid,
+												  tupleDescriptor, &isNull);
+		char *transactionName = TextDatumGetCString(transactionNameDatum);
 
-		if (!nameFound)
+		isTransactionInProgress = IsTransactionInProgress(activeTransactionNumberSet,
+														  transactionName);
+		if (isTransactionInProgress)
 		{
 			/*
-			 * baseName is not in subtractArray and thus included
-			 * in the difference.
+			 * Do not touch in progress transactions as we might mistakenly
+			 * commit a transaction that is actually in the process of
+			 * aborting or vice-versa.
 			 */
-			differenceList = lappend(differenceList, baseName);
+			continue;
 		}
-	}
 
-	pfree(subtractArray);
+		/*
+		 * Remove the transaction from the pending list such that only transactions
+		 * that need to be aborted remain at the end.
+		 */
+		hash_search(pendingTransactionSet, transactionName, HASH_REMOVE,
+					&isTransactionPending);
 
-	return differenceList;
-}
+		if (isTransactionPending)
```

Comment:
Hmm, I still observe `WARNING:  prepared transaction with identifier "citus_0_31126_358610_5477" does not exist` type of warning when pgbench continue to insert into a reference table and recovery runs in parallel for every second. 

Could it be possible that some transactions appear in `pendingTransactionList ` then commits and later `activeTransactionNumberList ` doesn't include that transaction which recovery would try to commit?

I think we should lock backends with `locklockdata()` at least getting for `pendingTransactionList ` and `activeTransactionNumberList `.   Looking into the perf output, that should be a very small amount of time. What do you think?

Reply:
> Could it be possible that some transactions appear in pendingTransactionList then commits and later activeTransactionNumberList doesn't include that transaction which recovery would try to commit?

The warnings occur during the recovery, not in the commit callback, right? I guess we could just ignore those.

> I think we should lock backends with locklockdata() at least getting for pendingTransactionList and activeTransactionNumberList. Looking into the perf output, that should be a very small amount of time. What do you think?

I'm a bit hesitant to do that, because we'd block both `assign_distributed_transaction_id` and COMMIT for the duration of a network round-trip (with possible failures, delays), instead of just blocking 2PC commit.

Reply:
Alternatively we could skip committing prepared transactions that are not in Q.

Reply:
Ok, the final changes make sense to me as well. Also, the tests run smoothly now